runtime/sema/type.go

```diff
@@ -4281,7 +4285,13 @@ func (t *CompositeType) initializeMemberResolvers() {
 			})
 
 		t.memberResolvers = withBuiltinMembers(t, memberResolvers)
-	})
+	}
+}
+
+func (t *CompositeType) ResolveMembers() {
+	if t.Members.Len() != len(t.GetMembers()) {
+		t.initializerMemberResolversFunc()()
+	}
```

Comment:
Why is this needed? Can the code calling this new `ResolveMembers` function just call `GetMember`, which will initialize members once/lazily?

Reply:
Hey @turbolent, thanks for looking into this :pray: 
I am puzzled myself, because this shouldn't be needed. I have just added this commit to showcase a strange behavior.
I have changed the `Test` contract, to have an `init` function, which expects an argument, e.g:
```cadence
pub contract Test {

    /// backend emulates a real network.
    ///
    pub let backend: AnyStruct{BlockchainBackend}

    init(backend: AnyStruct{BlockchainBackend}) {
        self.backend = backend
    }
    ...
}
```
Which resulted in tests failing with: 
```shell
Checking failed:
error: value of type `Test` has no member `assertEqual`
```
The reason behind this, appears to be a mismatch between the values returned by `compositeType.GetMembers()` and `compositeType.Members`, e.g.:
```go
compositeType.GetMembers():  [createAccount executeNextTransaction eventsOfType ScriptResult executeScript deployContract loadSnapshot Error account executeTransactions executeTransaction serviceAccount events assertError Matcher Transaction getType logs reset createSnapshot Account getAccount backend not isInstance beSucceeded beFailed beNil ResultStatus TransactionResult moveTime commitBlock addTransaction]
compositeType.Members:  [isInstance getType account backend executeScript createAccount getAccount addTransaction executeNextTransaction commitBlock executeTransaction executeTransactions deployContract logs serviceAccount events eventsOfType reset moveTime createSnapshot loadSnapshot not beSucceeded beFailed beNil assertError Matcher ResultStatus TransactionResult ScriptResult Error Account Transaction assert assertEqual fail readFile expect newMatcher equal beEmpty haveElementCount contain beGreaterThan beLessThan expectFailure]
```

The `Test` contract is partially written in a Cadence file, and then enriched with functions that are implemented natively. All this functions appear in the `compositeType.Members` field, but they are not included in `compositeType.GetMembers()`. That's why I added and manually call `compositeType.ResolveMembers()`.

I am guessing this has something to do with the addition of the `init()` function, I am not sure what's exactly happening though.

Reply:
Having a look on this now

Reply:
ah yeah, So what happens is:
- Test contract `.cdc` file is parsed and checked
- `Test` type gets initialized, and their members are also initialized and cached.
- However, at this point only the members written in cadence are available.
- Then the natively written members get "injected" to the `Test` type, but the member cache never gets updated.

Makes sense to re-instantiate the cache explicitly after injecting member 👍 
The only downside is, we won't be able to use these natively defined functions inside the `Test` contract `.cdc` file itself.

I think the longer-term approach is to add support for the `native` modifier in the checker as well (I think currently it is only supported in the parser). And then define those native functions also in Cadence, but with the `native` modifier.
e.g:
```cadence
struct Test {
    ...
    native assertEqual(...) {} 
}
```
The checker can declare these functions as members (only the signature/type), but the implementation will be provided natively. That way we won't need to manually inject the members to the `Test` type like we do now.

Anyway, we are still not here yet, and the current solution sounds good for now.

Reply:
That's a great explanation @SupunS :clap:

Reply:
I see. Yeah, it would be great to revert this workaround and properly declare all functions upfront in the Cadence code.

As discussed yesterday, adding support for `native` functions to the checker would be great, and should be very simple: if the function is declared as `native`, require the body to be empty and ignore it.

@SupunS Could you please open issues for adding this functionality and a tech-debt issue for reverting these changes here?

Reply:
Actually, I quickly implemented checking native functions in #2821. 

It would still be nice to have an issue that switches the test contract to make use of this feature, and reverts the workaround introduced here.

Reply:
Nice! Opened https://github.com/onflow/cadence/issues/2822.
I merged this PR for now, so the downstream PRs are not blocked by the refactor.